lib/src/strings.dart

```diff
@@ -94,38 +95,103 @@ String _tryYamlEncodeSingleQuoted(String string) {
+        ///
+        /// See https://yaml.org/spec/1.2.2/#64-empty-lines
+        if (current.trim().isEmpty) return previous + lineEnding + current;
+
+        var updated = current;
+
+        /// For consecutive non-empty lines, we add a `\n` with a space.
+        ///
+        /// We need to ensure the ACTUAL previous string was non-empty
+        if (trimmedSplit[index - 1].trim().isNotEmpty) {
+          /// We remove indent and check if the string starts with a space.
+          ///
+          /// Apply `\n` for `foo\nbar` but not `foo\n bar`.
+          if (!current.replaceFirst(indent, '').startsWith(' ')) {
+            updated = lineEnding + updated;
+          }
+        }
 
-  if (removedPortion.contains('\n')) {
-    result = '>+\n${' ' * indentation}';
-  } else {
-    result = '>-\n${' ' * indentation}';
+        /// Apply `\n` by default if no operation occurred
+        return previous + lineEnding + updated;
+      });
+    } else {
+      trimmedString = first;
+    }
   }
 
-  /// Duplicating the newline for folded strings preserves it in YAML.
-  /// Assumes the user did not try to account for windows documents by using
-  /// `\r\n` already
-  return result +
-      trimmedString.replaceAll('\n', lineEnding * 2 + ' ' * indentation) +
-      removedPortion;
+  return literalPrefix + trimmedString + stripped;
 }
 
 /// Generates a YAML-safe literal string.
 ///
 /// It is important that we ensure that [string] is free of unprintable
 /// characters by calling [_hasUnprintableCharacters] before invoking this
 /// function.
-String _tryYamlEncodeLiteral(
-    String string, int indentation, String lineEnding) {
-  final result = '|-\n$string';
+String _tryYamlEncodeLiteral(String string, int indentSize, String lineEnding) {
+  final (literalPrefix, indent, trimmedString, stripped) = _prepareBlockString(
+    '|',
+    string,
+    indentSize,
+    lineEnding,
+  );
+
+  /// Simplest block style.
+  /// * https://yaml.org/spec/1.2.2/#812-literal-style
+  return literalPrefix + indent + trimmedString + stripped;
+}
 
-  /// Assumes the user did not try to account for windows documents by using
-  /// `\r\n` already
-  return result.replaceAll('\n', lineEnding + ' ' * indentation);
+/// Prepares a string to be encoded either via [_tryYamlEncodeLiteral] or
+/// [_tryYamlEncodeFolded] when in [CollectionStyle.BLOCK].
+(
+  String literalPrefix,
+  String indent,
+  String trimmedString,
+  String strippedString,
+) _prepareBlockString(
+    String blockIndicator, String string, int indentSize, String lineEnding) {
+  final trimmed = string.trimRight();
+  final stripped = string.substring(trimmed.length);
+
+  final chompingIndicator =
+      string.endsWith('\n') || string.endsWith(' ') ? '+' : '-';
+
+  final indent = ' ' * indentSize;
+
+  return (
+    '$blockIndicator$chompingIndicator$lineEnding',
+    indent,
+
+    /// Assumes the user did not try to account for windows documents by using
+    /// `\r\n` already
```

Comment:
We don't have to state that. The user should never account for encoding details.

Sematically, a YAML document consists of YamlNodes, see package:yaml
If a YamlNode is a YamlScalar then it can have a `YamlScalar.value` that is a string.
If `value` is a string, then that is the string that must be returned when the document is parsed.

If the user actually gives us a string that contains `\r\n` we can't encode it in folded or literal mode. The only way to represent a `\r` character is to do quoted string `"\r\n"` -- am I wrong? Is it possible to string escaping in folded or literal strings? I imagine it's not possible, but I could be wrong.

Reply:
I'd suggest you take a look at:
https://github.com/dart-lang/yaml_edit/blob/main/lib/src/editor.dart#L564-L583

Essentially, if I use `YamlEditor.update(path, foo)` then `deepEquals(YamlEditor.parseAt(path), foo) == true`. And this MUST always be true (assuming that `deepEquals` ignores the styling and only looks at `.value`).
No matter what string I insert or what style I attempt to use, YamlEditor must fallback to use a less pretty style, if it can't represent a string in the style I chose.

For example, if I try to use special characters or `\0` null bytes or whatever, then this should still hold. But YamlEditor might have to fallback to quoted strings, even if I requested a Folded string style. The style is just a hint, not a requirement, because not all styles can be used for all values.